### PR TITLE
[ADD] beesdoo_print_label: new module to add product labels

### PR DESCRIPTION
**Pull request by flaenen**
This new module extends [product_print_category](https://github.com/grap/grap-odoo-incubator/tree/12.0/product_print_category). It adds a print category (Normal pricetag - 60x38mm) and the associated report to print 60x38 labels. 

Changing product.template fields does not currently indicate that the product label should be printed. It only works with the product.product fields. It is planned to adapt product_print_category.

[coopiteasy task](https://gestion.coopiteasy.be/web?#id=6483&view_type=form&model=project.task&menu_id=338&action=479)

### CODE REVIEW

**Comment by robinkeunen** [comment]
@flaenen can you rebase your work on branch 12.0 ? We just merged the CI configurations :-)

File: beesdoo_print_label/views/report_pricetag_normal.xml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                        </div>
+                        <div class="pricetag_left floating_box deposit">
+                            <t t-if="product_tmpl_id.total_deposit">
+                                <img src="/beesdoo_print_label/static/img/deposit.gif"/>
+                                <b><t t-esc="product_tmpl_id.total_deposit"/> <t t-esc="res_company.currency_id.symbol"/></b>
+                            </t>
+                        </div>
+                        <div class="pricetag_right floating_box">
+                            <div class="product_price">
+                                <t t-esc="'%0.2f %s' % (product_tmpl_id.total_with_vat, res_company.currency_id.symbol)"/>
```
Comment (robinkeunen) [inline beesdoo_print_label/views/report_pricetag_normal.xml]: I'm no QWeb expert but I think this should be handled like this :

```xml
<t t-esc="elem.total_amount" t-options="{'widget': 'monetary', 'display_currency': currency}"/>
```

File: beesdoo_print_label/data/product_print_category.xml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+            (4, ref('beesdoo_product.field_product_template__local_label')),
+            (4, ref('beesdoo_product.field_product_template__fair_label')),
+            (4, ref('beesdoo_product.field_product_template__origin_label')),
+            (4, ref('beesdoo_product.field_product_template__total_deposit')),
+            (4, ref('beesdoo_product.field_product_template__total_with_vat')),
+            (4, ref('beesdoo_product.field_product_template__display_weight')),
+            (4, ref('product.field_product_template__volume')),
+            (4, ref('beesdoo_product.field_product_template__display_unit')),
+            (4, ref('beesdoo_product.field_product_template__default_reference_unit')),
+            (4, ref('beesdoo_product.field_product_template__total_with_vat_by_unit')),]"/>
```
Comment (robinkeunen) [inline beesdoo_print_label/data/product_print_category.xml]: I would rather use a "6 instruction" here, something like 
```xml    
<field name="field_ids" eval="[(6, 0, [
            ref('product.field_product_template__name'),
            ref('beesdoo_product.field_product_template__main_seller_id'),
            ref('beesdoo_product.field_product_template__eco_label'),
            ref('beesdoo_product.field_product_template__local_label'),
            ref('beesdoo_product.field_product_template__fair_label'),
            ref('beesdoo_product.field_product_template__origin_label'),
            ref('beesdoo_product.field_product_template__total_deposit'),
            ref('beesdoo_product.field_product_template__total_with_vat'),
            ref('beesdoo_product.field_product_template__display_weight'),
            ref('product.field_product_template__volume'),
            ref('beesdoo_product.field_product_template__display_unit'),
            ref('beesdoo_product.field_product_template__default_reference_unit'),
            ref('beesdoo_product.field_product_template__total_with_vat_by_unit')])]"/>
```

File: beesdoo_print_label/views/report_pricetag_normal.xml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            <link href="/beesdoo_print_label/static/css/pricetag_base.scss" rel="stylesheet" type="text/scss"/>
+            <link href="/beesdoo_print_label/static/css/pricetag_normal_base.scss" rel="stylesheet" type="text/scss"/>
+            <link href="/beesdoo_print_label/static/css/pricetag_normal_normal.scss" rel="stylesheet" type="text/scss"/>
+        </xpath>
+    </template>
+
+    <template id="qweb_template_pricetag_normal">
+        <t t-call="web.basic_layout">
+            <t t-foreach="category_data['lines']" t-as="line">
+                <t t-foreach="line.quantity" t-as="q">
```
Comment (robinkeunen) [inline beesdoo_print_label/views/report_pricetag_normal.xml:30]: `q` is not used in later, did you actually mean to loop on a quantity ?

File: beesdoo_print_label/views/report_pricetag_normal.xml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                    <t t-set="product_tmpl_id" t-value="line.product_id.product_tmpl_id"/>
+                    <div class="normal_pricetag normal_pricetag_normal floating_box">
+                        <div class="product_name">
+                            <t t-esc="product_tmpl_id.name"/>
+                        </div>
+                        <div class="product_informations">
+                            <t t-if="product_tmpl_id.main_seller_id.name">
```
Comment (robinkeunen) [inline beesdoo_print_label/views/report_pricetag_normal.xml:40]: What happens if there is no `main_seller_id`? Shouldn't you rather test on `product_tmpl_id.main_seller_id`?

File: beesdoo_print_label/views/report_pricetag_normal.xml [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                                    <b><t t-esc="product_tmpl_id.volume"/> <t t-esc="product_tmpl_id.display_unit.name"/></b>
+                                </t>
+                                <t t-elif="product_tmpl_id.display_weight">Weight<br/>
+                                    <b><t t-esc="product_tmpl_id.display_weight"/> <t t-esc="product_tmpl_id.display_unit.name"/></b>
+                                </t>
+                            </div>
+                            <div class="product_price_per_uom_price floating_box">
+                                <t t-if="product_tmpl_id.total_with_vat_by_unit">
+                                    Price / <t t-esc="product_tmpl_id.default_reference_unit.name"/><br/>
+                                    <b><t t-esc="'%0.2f %s' % (product_tmpl_id.total_with_vat_by_unit, res_company.currency_id.symbol)"/></b>
```
Comment (robinkeunen) [inline beesdoo_print_label/views/report_pricetag_normal.xml]: Here as well.

**Review by robinkeunen — CHANGES_REQUESTED** [review]
Thank you very much for your help. I have a few comments.

As said in another comment : can you rebase your work on branch 12.0 ? We just merged the CI configurations :-)

File: beesdoo_print_label/views/report_pricetag_normal.xml (hunk 3 above)
Comment (flaenen) [reply beesdoo_print_label/views/report_pricetag_normal.xml:30]: line.quantity is the number of labels to print for a product.

File: beesdoo_print_label/views/report_pricetag_normal.xml (hunk 4 above)
Comment (flaenen) [reply beesdoo_print_label/views/report_pricetag_normal.xml:40]: Without main_seller_id, the recordset is empty and the condition is false.

Comment (robinkeunen) [reply beesdoo_print_label/views/report_pricetag_normal.xml:40]: 👌

File: beesdoo_print_label/views/report_pricetag_normal.xml (hunk 3 above)
Comment (robinkeunen) [reply beesdoo_print_label/views/report_pricetag_normal.xml:30]: Ok, I did not know you could do a foreach on an integer. 👌

**Review by robinkeunen — APPROVED** [review]
LGTM, just needs to pass the CI (black etc) 👌

**Comment by flaenen** [comment]
@robinkeunen Now it should pass...

**Comment by robinkeunen** [comment]
To fix the CI, you will need to add the repository in which the CI can find the product_print_category module to this file : https://github.com/beescoop/Obeesdoo/blob/12.0/oca_dependencies.txt

**Pull request merged**